engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/EnsoMultiValueTest.java

```
@@ -171,6 +172,93 @@ public void dataflowErrorPassingThroughMultiChecks() {
     assertEquals("(Error: (My_Error.Error 'msg'))", foo.toString());
   }
 
+  @Test
+  public void makeAbx13272() throws Exception {
+    var types =
+        """
+    from project.PrivateConversion import all
+    type A
+        A_Ctor x
+
+        a_method self = "A method"
+    type B
+        B_Ctor x
+
+        b_method self = "B method"
+
+    type X
+        X_Ctor x
+
+        x_method self = "X method"
+
+    make_abx -> A & B & X =
+        a = A.A_Ctor 1
+        # Relies on the hidden conversions
+        (a : A & B & X)
+    """;
+
+    var privateConversion =
```

question: What is "private" about these conversions? You import all of the from the `Types` module, so the `Main` module also sees all these conversions. How would the semantics be different if you just put all the conversions into `Types` module?

---

I think it's mostly our convention.

If you put them into `Types` module, then an import `from a.b.c.Types import all` would _accidentally_ bring these conversions into scope. However, **by design** these conversions should only be available in the place where they are used to construct multi values and _nowhere else_. Keeping them in a separate module makes sure an accidental import won't happen.

---

`private` in a sense that they aren't exposed to users of the `type`s.

> You import all of the from the Types module, 

That's correct.

> so the Main module also sees all these conversions

That's not true, I think.

---

> > so the Main module also sees all these conversions
> 
> That's not true, I think.

Can be easily proven in the test if you construct a 'plain' `A` value (just A constructor) and try to convert it to `B` - it should fail showing that the conversion is not available in the Main scope.

---

I think adding a check that verifies this assertion is worth adding to this test, as the mutli-value stuff only makes sense of the conversions are hidden - if the conversion is available the behaviour is (subtly) different.